### PR TITLE
`union`, `intersection` and `overlaps?` for `Range`s

### LINKED ISSUE #10148: Add `Range#clamp`
Range isn't `Comparable`, but it could very well define a `#clamp` method. The result would be the intersection of both ranges.

```cr
(-5..5).clamp(0..100)    # =>  0..5
(-5..5).clamp(-100..100) # => -5..5
(-5..5).clamp(..)        # => -5..5
(-5...5).clamp(-100...0) # => -5...0
```

This method could also be called `#intersect` or `#&`, but IMO `#clamp` fit's best because it closely resembles `Comparable#clamp`.

There are a few cases where the expected results may be up for discussion:

* Mixing exclusive and inclusive ranges (`(-5..5).clamp(-10...2)` and `(-5...5).clamp(-10..2)`) should probably have the same result as if both ranges were exclusive.
* If both ranges are disparate (`(-5..5).clamp(10..100)`), the result should probably be an empty range. This can be expressed as an exclusive range with identical begin and end (`x...x`). But what should `x` be? Probably best to use the begin of the clamped range (`-5...-5`).
* If the direction of the ranges is different (`-5..5.clamp(10..-10)`), we could try to sort things out, but I suppose it's probably better to just treat them as disparate and return an empty range.

### LINKED ISSUE #14487: Add a method to `Range` to detect overlap/intersection with another `Range`
The `Range` type doesn't currently tell you if it overlaps with another instance, but it's a useful thing to check for. For example, a scheduling app may need to detect whether a given time period conflicts with any existing meetings/events:

```crystal
existing_events.any?(&.overlaps?(range))
```

### LINKED ISSUE #14488: Union and intersection of `Range`s
Combined with #14487, it would be useful to have `Range` methods that, when given another range, would return a new range that contains the union or intersection of both ranges.

For example:

```crystal
a = 1..5
b = 3..7

a.union(b)        # => 1..7
a.intersection(b) # => 3..5
```

If there is no overlap between them, I'm not 100% sure what should happen but I feel like returning `nil` makes sense.

```crystal
a = 1..5
c = 10..20

a.union(c)        # => nil
a.intersection(c) # => nil
```

There are probably some other edge cases to work out, as well, such as:
- what happens when ranges are infinite in one direction?
  - should the `union` be infinite in that direction?
  - should the `intersection` use whichever one has a finite value?
- for ranges like `1...10` and `10..20` where there's nooverlap, but they _are_ adjacent:
  - should the union be `1..20` — the union of adjacent ranges providing continuity
  - should the intersection be `nil` (or whatever the non-overlap result is deemed to be)?
- should `(1...10).union(1..10)` be `1..10`?
- should `(1...10).intersection(1..10)` be `1...10`?
- any other edge cases?

### PR DESCRIPTION
**Pull request by CTC97**
Apologies for the earlier PR from my forked master branch.

Based on the request from @jgaskins. Behaviors below:

```
a = 1..5
b = 3..7
a.union(b)        # => 1..7
a.intersection(b) # => 3..5

c = 1..10
d = 1...15
c.union(d)        # => 1...15
c.intersection(d) # => 1..10

e = 1...10
f = 1..15
e.union(f)        # => 1..15
e.intersection(f) # => 1...10

g = 1..5
h = 10..20
g.union(h)        # => 0..0
g.intersection(h) # => 0..0

(1...10).union(1..10)        # => 1..10
(1...10).intersection(1..10) # => 1...10

(1...10).union(10..20) # => 1..20

(1..5).union(5..10) # => 1..10

(1...10).intersection(7..12) # => 7...10
(2..10).intersection(0..8)   # => 2..8

(1...10).overlaps?(10..11)   # => false
(1..10).overlaps?(5..9) # => true
```

Fixes #14487
Fixes #14488
Fixes #10148

---

P.S., very excited to be making a small contribution to a lovely language. 

Looking forward to review!

### CODE REVIEW

**Comment by jgaskins** [comment]
This looks like a great start. A couple areas of focus:

- Ranges can be `nil` on either end
- It would be awesome to see some tests — this functionality can be super confusing and, honestly, I wouldn't be able to tell whether any implementation was correct when all edge cases are considered

File: src/range.cr [hunk 1]
```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
...
+  # Returns the union of this range, and another.
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (1..5).union(5..10)   # => 1..10
+  # (1...10).union(1..10) # => 1..10
+  # ```
+  def union(other : Range)
+    if self.end < other.begin || other.end < self.begin
+      return 0..0
```
Comment (jgaskins) [inline src/range.cr:264]: I'm not sure we can get away with hardcoding a `Range(Int32, Int32)` specifically. My own use case for this functionality is actually `Range(Time, Time)`, so I'd get compile-time errors with this return value.

Comment (CTC97) [reply src/range.cr:264]: Good catch, and thank you for the quick review! 

I'll get on these changes asap! ✌️⭐

**Comment by Blacksmoke16** [comment]
Some specs would be a good idea as well :+1:.

**Comment by jgaskins** [comment]
Tests *and* specs? In this economy?

File: src/range.cr (hunk 1 above)
Comment (HertzDevil) [reply src/range.cr:264]: Using any of the four endpoints here instead of `0` should probably work too

Comment (CTC97) [reply src/range.cr:264]: To me nil seems like a more logical return value than using any of the existing endpoints which might be confusing. 

Totally open to your reasoning though!

**Comment by CTC97** [comment]
Uploaded some specs and went with the nil return as originally suggested in @jgaskins' issue. 

On error handling - should I be adding any type handling in the functions themselves or is an error message like below sufficient? 

```
(1..5).union(time_range)

---

In src/range.cr:263:25

 263 | if self.end < other.begin || other.end < self.begin
                           ^----
Error: expected argument #1 to 'Int32#<' to be Float32, Float64, Int128, Int16, Int32, Int64, Int8, Number, UInt128, UInt16, UInt32, UInt64 or UInt8, not Time
``` 

or 

```
(time_range).union(1..5)

---

In src/range.cr:263:25

 263 | if self.end < other.begin || other.end < self.begin
                           ^----
Error: expected argument #1 to 'Time#<' to be Time, not Int32
```

**Comment by straight-shoota** [comment]
I don't think we can do much about improving these error message anyways. That would require knowing for which types the comparison operator is defined, and we cannot know this. We mustn't restrict it to identical types because there are many examples where different typs are still comparable.

**Comment by straight-shoota** [comment]
Adjacency can only work for types with (meaningfully) discrete values. This should be indicated by `#succ` and `#pred` methods defined on the item type. `Range#each` (and `#reverse_each`) are based on that.
Technically, most (all?) comparable types have discrete steps in some form, but they're not always meaningful. `Float#next_float` and `time + 1.nanosecond` would return adjacent values, but those are defined in the representation and not very useful in practice.

So for types that define `#succ` or `#pred`, we could discover adjacency. For all others, to form a union the ranges must at least touch. This can match exclusive ranges though: `(1.0...5.0).union(5.0..10.0).should eq(1.0..10.0)`.

File: spec/std/range_spec.cr [hunk 2]
```diff
@@ -80,23 +80,84 @@ describe "Range" do
...
+    # TODO: This test will fail - 
+    #   In an integer setting this case is doable. We can simply check adjaceny with +-1.
+    #   However, how do we define an adjacent range for other data types? 
+    #     Examples of adjacency ambiguity:
+    #       - Floats: Is 4.1's upper adjacent value 4.2 or 4.11, etc.?
+    #       - Times: Is a time's adjacent value +1 second or +1 hour, etc.?
+    # 
+    #   How do we define adjacency?
+    #   
+    # (1..5).union(6..10).should eq(1..10)      # Adjacent ranges
```
Comment (ysbaddaden) [inline spec/std/range_spec.cr:95]: As per the [Range documentation](https://crystal-lang.org/api/1.14.0/Range.html) you'd use `#succ` and `#pred`:

> Ranges typically involve integers, but can be created using arbitrary objects as long as they define succ (or pred for [#reverse_each](https://crystal-lang.org/api/1.14.0/Range.html#reverse_each%28%26%29%3ANil-instance-method)), to get the next element in the range, and < and [#==](https://crystal-lang.org/api/1.14.0/Range.html#%3D%3D%28other%3ARange%29-instance-method), to know when the range reached the end:

File: src/range.cr (hunk 1 above)
Comment (ysbaddaden) [reply src/range.cr]: Think the intersection of a list returning a nil vs an empty list: which feels more logical? Which is less of a burden?

Returning `nil` means that the methods _always_ return a nilable type (`Range(B, E) | Nil`) and we must _always_ check that whether the value is nil or not.

Returning an empty range means we always return a Range(B, E), which would act as a NOOP object.

Comment (ysbaddaden) [reply src/range.cr]: BTW: `0..0` wasn't an empty range, it's a range with a single value (starts at 0 and terminates at 0 inclusive):

```crystal
(0..0).to_a # => [0]
(0...0).to_a # => []
```

**Comment by ysbaddaden** [comment]
What about some edge cases? For example empty ranges:

``crystal
(1...1).intersect(0..10)
```

File: src/range.cr [hunk 3]
```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
...
+    if self.end < other.begin || other.end < self.begin
+      return nil
+    end
+
+    larger = (self.end > other.end)
+    if (larger && self.excludes_end?) || (!larger && other.excludes_end?)
+      return (self.begin < other.begin ? self.begin : other.begin)...(self.end > other.end ? self.end : other.end)
+    end
+
+    (self.begin < other.begin ? self.begin : other.begin)..(self.end > other.end ? self.end : other.end)
```
[the remark this answers is not in the document]
Comment (CTC97) [reply src/range.cr]: This is great, thanks for the suggestion! Just pushed a change to implement this.

**Comment by CTC97** [comment]
> What about some edge cases? For example empty ranges:
> 
> ```crystal
> (1...1).intersect(0..10) #=> empty
> ```

I can add a check for one of the ranges being empty (i.e. non-inclusive ranges where the ends are equal), but is it valid to return 1...1, which is an accurate representation of the intersection. (...is it accurate?)

File: spec/std/range_spec.cr (hunk 2 above)
Comment (CTC97) [reply spec/std/range_spec.cr:95]: > As per the [Range documentation](https://crystal-lang.org/api/1.14.0/Range.html) you'd use `#succ` and `#pred`:
> 
> > Ranges typically involve integers, but can be created using arbitrary objects as long as they define succ (or pred for [#reverse_each](https://crystal-lang.org/api/1.14.0/Range.html#reverse_each%28%26%29%3ANil-instance-method)), to get the next element in the range, and < and [#==](https://crystal-lang.org/api/1.14.0/Range.html#%3D%3D%28other%3ARange%29-instance-method), to know when the range reached the end:

Making this change this weekend. ✨

File: src/range.cr [hunk 4]
```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
...
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (1..5).union(5..10)   # => 1..10
+  # (1...10).union(1..10) # => 1..10
+  # ```
+  def union(other : Range)
+    if self.end < other.begin || other.end < self.begin
+      return nil
+    end
```
Comment (Sija) [inline src/range.cr:265]: Might be shortened a bit:

```suggestion
    return if self.end < other.begin || other.end < self.begin
```

ditto below

File: src/range.cr [hunk 5]
```diff
@@ -252,6 +252,52 @@ struct Range(B, E)
...
 
+  # Returns the union of this range, and another.
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (1..5).union(5..10)   # => 1..10
+  # (1...10).union(1..10) # => 1..10
+  # ```
+  def union(other : Range)
+    return nil if self.end < other.begin || other.end < self.begin
```
Comment (Sija) [inline src/range.cr:263]:
```suggestion
    return if self.end < other.begin || other.end < self.begin
```

File: src/range.cr [hunk 6]
```diff
@@ -252,6 +252,52 @@ struct Range(B, E)
...
+
+  # Returns the intersection of this range, and another.
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (2..10).intersection(0..8)   # => 2..8
+  # (1...10).intersection(7..12) # => 7...10
+  # ```
+  def intersection(other : Range)
+    return nil if self.end < other.begin || other.end < self.begin
```
Comment (Sija) [inline src/range.cr:281]:
```suggestion
    return if self.end < other.begin || other.end < self.begin
```

**Review by Sija — COMMENTED** [review]
`nil` is implicit with no return value.

**Comment by CTC97** [comment]
Apologies for the untimely updates here - been a bit of a hectic few weeks, but will be jumping back into this asap! I have not forgotten! :sweat_smile:

**Comment by CTC97** [comment]
I want to do something like:

```
return if self.end < other.begin && (!self.end.responds_to?(:succ) || self.end.succ != other.begin)
return if other.end < self.begin && (!other.end.responds_to?(:succ) || other.end.succ != self.begin)
```

for the ```union``` method but am running into issues on compile. How can I dynamically check for :succ presence and not run into compiling errors?

I think it makes sense to avoid explicitly defining a list of types that support it within the function and checking against that.

File: src/range.cr [hunk 7]
```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
...
+    larger = (self.end > other.end)
+    Range.new(
+      self.begin < other.begin ? self.begin : other.begin,
+      self.end > other.end ? self.end : other.end,
+      exclusive: (larger && self.excludes_end?) || (!larger && other.excludes_end?),
+    )
+  end
+
+  # Returns the intersection of this range, and another.
+  # Returns 0..0 if there is no overlap.
```
Comment (straight-shoota) [inline src/range.cr:278]: **issue:** The documentation needs to adapt to the change in behaviour:

```suggestion
  # Returns `nil` if there is no overlap.
```

File: src/range.cr [hunk 8]
```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
...
+  # ```
+  def intersection(other : Range)
+    return if self.end < other.begin || other.end < self.begin
+
+    larger = (self.end > other.end)
+    Range.new(
+      self.begin > other.begin ? self.begin : other.begin,
+      self.end < other.end ? self.end : other.end,
+      exclusive:(!larger && self.excludes_end?) || (larger && other.excludes_end?),
+    )
```
Comment (straight-shoota) [inline src/range.cr:292]: **issue:** We're completely missing tests for `exclusive` ranges.
And it looks like the current implementation might actually not be correct. It seems to take `excludes_end?` of ther smaller of `self` and `other`, but shouldn't it be the larger of both?

I believe this could be implemented in a way that's simpler and more straightforward to reason about:

```suggestion
    larger = (self.end > other.end)
    Range.new(
      Math.max(self.begin, other.begin),
      Math.min(self.end, other.end),
      exclusive:({self, other}.max_by(&.end).excludes_end?,
    )
```

Ditto for `#union`.

File: src/range.cr [hunk 9]
```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
...
+    e_s : E = self.end
+    e_o : E = other.end
+
+    return if self.end < other.begin && (!e_s.responds_to?(:succ) || e_s.succ != other.begin)
+    return if other.end < self.begin && (!e_o.responds_to?(:succ) || e_o.succ != self.begin)
+
+    Range.new(
+      Math.min(self.begin, other.begin),
+      Math.max(self.end, other.end),
+      exclusive: ({self, other}.max_by(&.end).excludes_end?),
```
Comment (Sija) [inline src/range.cr:272]: Redundant parenthesis.

```suggestion
      exclusive: {self, other}.max_by(&.end).excludes_end?,
```

File: src/range.cr [hunk 10]
```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
...
+  # (2..10).intersection(0..8)   # => 2..8
+  # (1...10).intersection(7..12) # => 7...10
+  # ```
+  def intersection(other : Range)
+    return if self.end < other.begin || other.end < self.begin
+
+    Range.new(
+      Math.max(self.begin, other.begin),
+      Math.min(self.end, other.end),
+      exclusive: ({self, other}.max_by(&.end).excludes_end?),
```
Comment (Sija) [inline src/range.cr:289]: ditto

```suggestion
      exclusive: {self, other}.max_by(&.end).excludes_end?,
```

**Comment by ysbaddaden** [comment]
> I can add a check for one of the ranges being empty (i.e. non-inclusive ranges where the ends are equal), but is it valid to return 1...1, which is an accurate representation of the intersection. (...is it accurate?)

Well, `Range(1...1)` is empty, so I don't expect it to intersect with anything but another empty range. IMO it should return an empty range. Hence the suggestions to return an empty (whatever the chosen values) instead of `nil`. The question is whether returning whatever values can have some edge effects?

**Comment by CTC97** [comment]
> Well, 1...1 is empty, so I don't expect it to intersect with anything but another empty range. IMO it should return an empty range. Hence the suggestions to return an empty (whatever the chosen values) instead of nil. The question is whether returning whatever values can have some edge effects?

I could imagine some edge effects, but only if the end user was comparing data to either the beginning or end of the empty range without considering context of the full object (i.e., both ends).

I don't think it's unreasonable to expect users to handle that sort of return if their use case could result that way, but I'm open to other thoughts.

**Comment by ysbaddaden** [comment]
It's not that `nil` is an invalid value, but it's cumbersome. Anytime you intersect you MUST deal with a nilable, when an empty range would be a transparent NOOP.

**Comment by CTC97** [comment]
Coming back to this - what do people think is a reasonable empty range to return in the edge cases discussed above? I'm not sure if we've reached a solid conclusion.

**Comment by CTC97** [comment]
I implemented empty ranges with the logic @straight-shoota describes in #10148, which describes a `#clamp` function (which is interchangeable with `#intersect`):

> If both ranges are disparate ((-5..5).clamp(10..100)), the result should probably be an empty range. This can be expressed as an exclusive range with identical begin and end (x...x). But what should x be? Probably best to use the begin of the clamped range (-5...-5).

I think this is a good resolution as it logically makes sense and won't require the user to handle nil.

I've also added #10148 to this PR, but I think there's a case to be made for giving `#intersection` an alias `#clamp`, which users might find intuitive.

File: spec/std/range_spec.cr [hunk 11]
```diff
@@ -79,6 +79,89 @@ describe "Range" do
     (1...5).includes?(5).should be_false
   end
 
+  it "does union" do
+    # Int ranges
+    (1..5).union(3..7).should eq(1..7)    # Overlapping integer ranges
+    (1..5).union(6..10).should eq(1..10)  # Adjacent ranges
+    (1..5).union(10..15).should eq(1...1) # Disjoint integer ranges
```
Comment (Sija) [inline spec/std/range_spec.cr:86]: I believe this relays the intention better:

```suggestion
    (1..5).union(10..15).should be_empty # Disjoint integer ranges
```

ditto elsewhere

Comment (CTC97) [reply spec/std/range_spec.cr:86]: Using `be_empty` here results in an issue with one of the range macros:

```
Showing last frame. Use --error-trace for full trace.

There was a problem expanding macro 'macro_4849671408'

Code in src/range.cr:120:5

 120 | {% if E == Nil %}
       ^
Called macro defined in src/range.cr:120:5

 120 | {% if E == Nil %}

Which expanded to:

 > 3 | while end_value.nil? || current < end_value
 > 4 |   yield current
 > 5 |   current = current.succ
                           ^---
Error: undefined method 'succ' for Float64
```
`.empty?` must leverage this macro? I'm not sure, but I also don't think we necessarily need to use `be_empty` to in the spec.

**Review by straight-shoota — DISMISSED** [review]

File: spec/std/range_spec.cr [hunk 12]
```diff
@@ -79,6 +79,89 @@ describe "Range" do
...
+  it "does intersection" do
+    # Int ranges
+    (1..5).intersection(3..7).should eq(3..5)   # Overlapping integer ranges
+    (1..5).intersection(6..10).should eq(1...1) # Non-overlapping integer ranges
+    (1..5).intersection(5..10).should eq(5..5)
+    (1...1).intersection(0..10).should eq(1...1)
+
+    # Float ranges
+    (1.0..5.5).intersection(3.2..7.8).should eq(3.2..5.5)  # Overlapping float ranges
+    (1.0..2.5).intersection(3.0..4.0).should eq(1.0...1.0) # Non-overlapping float ranges
```
Comment (jgaskins) [inline spec/std/range_spec.cr:122]: Added a test case:
- adjacent float ranges with `self` excluding `other.begin`

> [!WARNING]
> This test fails

```suggestion
    # Float ranges
    (1.0..5.5).intersection(3.2..7.8).should eq(3.2..5.5)   # Overlapping float ranges
    (1.0...3.0).intersection(3.0..4.0).should eq(1.0...1.0) # Adjacent float ranges
    (1.0..2.5).intersection(3.0..4.0).should eq(1.0...1.0)  # Non-overlapping float ranges
```

File: spec/std/range_spec.cr [hunk 13]
```diff
@@ -79,6 +79,89 @@ describe "Range" do
...
+    (1...5).intersection(3...7).should eq(3...5)             # Overlapping exclusive integer ranges
+    (1.0...5.5).intersection(3.2...7.8).should eq(3.2...5.5) # Overlapping exclusive float ranges
+    ('a'...'e').intersection('c'...'g').should eq('c'...'e') # Overlapping exclusive string ranges
+    (t1...t3).intersection(t2...t4).should eq(t2...t3)       # Overlapping exclusive time ranges
+  end
+
+  it "overlaps?" do
+    # Int ranges
+    (1..5).overlaps?(3..7).should eq(true)   # Overlapping integer ranges
+    (1..5).overlaps?(6..10).should eq(false) # Non-overlapping integer ranges
```
Comment (jgaskins) [inline spec/std/range_spec.cr:146]: Another quick check for dependency on order of ranges and added a range with very slight overlap (`1..6` and `6..10`).

> [!WARNING]
> The test for `1..6` and `6..10` fails

```suggestion
    # Int ranges
    (1..5).overlaps?(3..7).should eq(true)   # Overlapping integer ranges
    (3..7).overlaps?(1..5).should eq(true)   # Overlapping integer ranges
    (1..6).overlaps?(6..10).should eq(true)  # Overlapping integer ranges
    (1..5).overlaps?(6..10).should eq(false) # Non-overlapping integer ranges
    (6..10).overlaps?(1..5).should eq(false) # Non-overlapping integer ranges
```

File: src/range.cr [hunk 14]
```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
...
+
+  # Returns the intersection of this range, and another.
+  # Returns `self.begin...self.begin` if there is no overlap.
+  #
+  # ```
+  # (2..10).intersection(0..8)   # => 2..8
+  # (1...10).intersection(7..12) # => 7...10
+  # ```
+  def intersection(other : Range)
+    return Range.new(self.begin, self.begin, exclusive: true) if self.end < other.begin || other.end < self.begin
```
Comment (jgaskins) [inline src/range.cr:284]: An intersection is empty when there is no overlap.

This refactor is actually how I uncovered the bug in `overlaps?` for the very slightly overlapping integer ranges. I totally would've missed it otherwise.

```suggestion
    return Range.new(self.begin, self.begin, exclusive: true) unless overlaps?(other)
```

**Review by jgaskins — CHANGES_REQUESTED** [review]
This looks fantastic! I have some suggestions for a few new test cases, nearly all of which passed — I'm suggesting them mainly to cover regression.

If I counted correctly, 3 of my suggested assertions do fail, but 2 are pedantic failures (they return `other.begin...other.begin` instead of `self.begin...self.begin`). @straight-shoota mentioned the exact return value being meaningful here and, although I can't _imagine_ a scenario where it would cause issues, I agree with him because we can't assume others' use cases will be like our own.

The less pedantic, more realistic failure is that `(1..6).overlaps?(6..10)` returns `false` instead of `true` because they both have the value `6` in common.

File: spec/std/range_spec.cr [hunk 15]
```diff
@@ -79,6 +79,89 @@ describe "Range" do
...
+    (1..5).union(10..15).should eq(1...1) # Disjoint integer ranges
+
+    # Float ranges
+    (1.0..5.5).union(3.2..7.8).should eq(1.0..7.8)  # Overlapping float ranges
+    (1.0..2.5).union(3.0..4.0).should eq(1.0...1.0) # Non-overlapping float ranges
+
+    # String ranges
+    ('a'..'e').union('c'..'g').should eq('a'..'g')  # Overlapping string ranges
+    ('a'..'c').union('e'..'f').should eq('a'...'a') # Non-overlapping string ranges
+    ('a'..'c').union('d'..'f').should eq('a'..'f')  # Adjacent string ranges
```
[the remark this answers is not in the document]
Comment (jgaskins) [reply spec/std/range_spec.cr:95]: … speaking of being pedantic 😄

File: src/range.cr [hunk 16]
```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
     {% end %}
   end
 
+  # Returns the union of this range, and another.
+  # Returns `self.begin...self.begin` if there is no overlap.
```
Comment (ysbaddaden) [inline src/range.cr:256]: **Suggestion (not blocking)**: it would be nice to spell out that it's returning an _empty range_ that is `self.begin...self.begin` when there is no overlap.

Same for `#intersection`.

**Review by ysbaddaden — COMMENTED** [review]
Very nice! Thanks for coping with us while refining this tricky feature :)

File: src/range.cr [hunk 17]
```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
...
+  end
+
+  # Returns `true` if this range overlaps with another range.
+  #
+  # ```
+  # (1..10).overlaps?(5..9)    # => true
+  # (1...10).overlaps?(10..11) # => false
+  # ```
+  def overlaps?(other : Range) : Bool
+    !(self.end <= other.begin || other.end <= self.begin)
```
Comment (Sija) [inline src/range.cr:300]: I'd suggest avoiding negation to make the code more clear:

```suggestion
    self.end >= other.begin || other.end >= self.begin
```

**Comment by straight-shoota** [comment]
Specs are failing.
And we're missing specs for endless and beginless ranges (probably also the implementation doesn't handle them correctly yet).

A couple of test cases for the clamp behaviour: https://github.com/straight-shoota/crystal/blob/27c3ee98d22c48d426192c75e7c2656ef4ddb53c/spec/std/range_spec.cr#L809-L829

**Comment by straight-shoota** [comment]
@CTC97 Are you up for completing this PR?
Or would you prefer someone else to take it over the finish line?

**Comment by CTC97** [comment]
@straight-shoota Happy to let someone finish this one - sorry for falling behind on this one guys!

**Comment by straight-shoota** [comment]
Comparison with the specs from https://github.com/crystal-lang/crystal/pull/10300 reveals this PR misses or only partially covers some scenarios. For example, open-ended ranges or ranges abutting at an exclusive end.
There might be a couple of edge cases on top of that.

That's just to say it'll require a bit more work to finalize this.
